### PR TITLE
Trie optimizations

### PR DESCRIPTION
**Pull request by debris**
changes:
- triedb seek and descend is not recursive
- optimized memorydb `insert`, `remove` and `emplace`
- optimized hashdb `keys()`

all trie insert benchmarks are about 10% faster

### CODE REVIEW

File: util/src/memorydb.rs [hunk 1]
```diff
@@ -200,16 +203,17 @@ impl HashDB for MemoryDB {
 			return SHA3_NULL_RLP.clone();
 		}
 		let key = value.sha3();
-		if match self.data.get_mut(&key) {
-			Some(&mut (ref mut old_value, ref mut rc @ -0x80000000i32 ... 0)) => {
-				*old_value = DBValue::from_slice(value);
+		match self.data.entry(key) {
+			Entry::Occupied(mut entry) => {
+				let &mut (ref mut old_value, ref mut rc) = entry.get_mut();
+				if *rc >= -0x80000000i32 && *rc <= 0 {
```
Comment (debris) [inline util/src/memorydb.rs]: `rc >= -0x80000000i32` is probably not needed, right @rphmeier ?

File: util/src/memorydb.rs [hunk 2]
```diff
@@ -181,7 +181,10 @@ impl HashDB for MemoryDB {
 	}
 
 	fn keys(&self) -> HashMap<H256, i32> {
-		self.data.iter().filter_map(|(k, v)| if v.1 != 0 {Some((k.clone(), v.1))} else {None}).collect()
+		self.data.iter()
```
Comment (arkpar) [inline util/src/memorydb.rs]: Why remove a `filter_map`? Clippy produces a warning on `.filter().map()` sequence

File: util/src/trie/triedb.rs [hunk 3]
```diff
@@ -217,59 +217,76 @@ impl<'a> TrieDBIterator<'a> {
...
-						status: Status::At,
-						node: node.clone().into(),
-					});
-				} else {
-					self.trail.push(Crumb {
-						status: Status::Exiting,
-						node: node.clone().into(),
-					});
-				}
+	fn seek_descend<'key>(&mut self, mut node_data: DBValue, mut key: NibbleSlice<'key>) -> super::Result<()> {
```
Comment (arkpar) [inline util/src/trie/triedb.rs]: Should be renamed to just `seek` since it is not recursive anymore.

File: util/src/trie/triedb.rs [hunk 4]
```diff
@@ -217,59 +217,76 @@ impl<'a> TrieDBIterator<'a> {
...
+		enum Step {
+			Descend(DBValue, usize),
+			Return,
```
Comment (arkpar) [inline util/src/trie/triedb.rs]: This could be just a `return` statement

File: util/src/memorydb.rs (hunk 2 above)
Comment (debris) [reply util/src/memorydb.rs]: ah, ok

I just found it more readable than this long one-liner

**Comment by debris** [comment]
@arkpar fixed

File: util/src/memorydb.rs (hunk 2 above)
Comment (debris) [reply util/src/memorydb.rs]: fixed

**Comment by debris** [comment]
js failure is unrelated

**Pull request merged**